File: lib/src/krpc/encode.rs
use std::collections::HashMap;
use std::hash::{BuildHasher, Hash};
use std::io::{Cursor, Write};

use color_eyre::eyre;
use prost::bytes::{Buf, BufMut};
use prost::Message;
use varint_rs::{VarintReader, VarintWriter};

use super::krpc::schema::{Dictionary, DictionaryEntry, List, Status, Tuple};
use super::{
    CelestialBody, Decoupler, Editor, EditorParts, EditorShip, LaunchClamp, Module, Orbit,
    PFDecoupler, Part, Parts, RODecoupler, ReferenceFrame, Resource, Resources, Vessel,
};

pub trait DecodeValue: Sized {
    fn decode_value(buf: &[u8]) -> eyre::Result<Self>;
}

pub trait EncodeValue: Sized {
    fn encode_value(self) -> eyre::Result<Vec<u8>>;
}

impl<K: DecodeValue + Hash + Eq, V: DecodeValue, S: BuildHasher + Default> DecodeValue
    for HashMap<K, V, S>
{
    fn decode_value(buf: &[u8]) -> eyre::Result<Self> {
        let mut map = HashMap::default();
        let dict = Dictionary::decode(&mut &*buf)?;
        for entry in dict.entries {
            let key = K::decode_value(&entry.key)?;
            let value = V::decode_value(&entry.value)?;
            map.insert(key, value);
        }
        Ok(map)
    }
}

impl<K: EncodeValue + Hash + Eq, V: EncodeValue, S: BuildHasher + Default> EncodeValue
    for HashMap<K, V, S>
{
    fn encode_value(self) -> eyre::Result<Vec<u8>> {
        let mut entries = vec![];
        for (key, value) in self {
            let key = key.encode_value()?;
            let value = value.encode_value()?;
            entries.push(DictionaryEntry { key, value });
        }
        Ok(Dictionary { entries }.encode_to_vec())
    }
}

impl<T: DecodeValue> DecodeValue for Vec<T> {
    fn decode_value(buf: &[u8]) -> eyre::Result<Self> {
        let mut v = vec![];
        let list = List::decode(&mut &*buf)?;
        for item in list.items {
            v.push(T::decode_value(&item)?);
        }
        Ok(v)
    }
}

impl<T: EncodeValue> EncodeValue for Vec<T> {
    fn encode_value(self) -> eyre::Result<Vec<u8>> {
        let mut v = vec![];
        for item in self {
            v.push(item.encode_value()?);
        }
        Ok(List { items: v }.encode_to_vec())
    }
}

macro_rules! encode_tuples {
    ($($($x:ident),* => $($ix:literal: $i:ident),*);*) => {
	$(
	    impl<$($x: EncodeValue),*> EncodeValue for ($($x),*,) {
		fn encode_value(self) -> eyre::Result<Vec<u8>> {
		    let mut v = vec![];
		    let ($($i,)*) = self;
		    $(v.push($i.encode_value()?);)*
		    Ok(Tuple { items: v }.encode_to_vec())
		}
	    }

	    impl<$($x: DecodeValue),*> DecodeValue for ($($x),*,) {
		fn decode_value(buf: &[u8]) -> eyre::Result<Self> {
		    let tuple = Tuple::decode(&mut &*buf)?;
		    let tuple = ($(
			$x::decode_value(&tuple.items[$ix])?,
		    )*);
		    Ok(tuple)
		}
	    }
	)*
    }
}

encode_tuples![
    T => 0: a;
    T, U => 0: a, 1: b;
    T, U, V => 0: a, 1: b, 2: c;
    T, U, V, W => 0: a, 1: b, 2: c, 3: d;
    T, U, V, W, X => 0: a, 1: b, 2: c, 3: d, 4: e
];

impl DecodeValue for String {
    fn decode_value(buf: &[u8]) -> eyre::Result<Self> {
        let mut cur = Cursor::new(buf);
        let _ = cur.read_u64_varint();
        let pos = cur.position() as usize;
        Ok(String::from_utf8(buf[pos..].to_vec())?)
    }
}

impl EncodeValue for String {
    fn encode_value(self) -> eyre::Result<Vec<u8>> {
        let mut v = vec![];
        v.write_u64_varint(self.len() as u64)?;
        v.write_all(self.as_bytes())?;
        Ok(v)
    }
}

impl<T: DecodeValue> DecodeValue for Option<T> {
    fn decode_value(buf: &[u8]) -> eyre::Result<Self> {
        if buf == [0u8] {
            Ok(None)
        } else {
            Ok(Some(T::decode_value(buf)?))
        }
    }
}

impl<T: EncodeValue> EncodeValue for Option<T> {
    fn encode_value(self) -> eyre::Result<Vec<u8>> {
        match self {
            Some(v) => v.encode_value(),
            None => Ok(vec![0u8]),
        }
    }
}

impl DecodeValue for f32 {
    fn decode_value(mut buf: &[u8]) -> eyre::Result<Self> {
        Ok(buf.get_f32_le())
    }
}

impl EncodeValue for f32 {
    fn encode_value(self) -> eyre::Result<Vec<u8>> {
        let mut v = vec![];
        v.put_f32_le(self);
        Ok(v)
    }
}

impl DecodeValue for f64 {
    fn decode_value(mut buf: &[u8]) -> eyre::Result<Self> {
        Ok(buf.get_f64_le())
    }
}

impl EncodeValue for f64 {
    fn encode_value(self) -> eyre::Result<Vec<u8>> {
        let mut v = vec![];
        v.put_f64_le(self);
        Ok(v)
    }
}

impl DecodeValue for i32 {
    fn decode_value(mut buf: &[u8]) -> eyre::Result<Self> {
        Ok(buf.read_i32_varint()?)
    }
}

impl EncodeValue for i32 {
    fn encode_value(self) -> eyre::Result<Vec<u8>> {
        let mut v = vec![];
        v.write_i32_varint(self)?;
        Ok(v)
    }
}

impl DecodeValue for u32 {
    fn decode_value(mut buf: &[u8]) -> eyre::Result<Self> {
        Ok(buf.read_u32_varint()?)
    }
}

impl EncodeValue for u32 {
    fn encode_value(self) -> eyre::Result<Vec<u8>> {
        let mut v = vec![];
        v.write_u32_varint(self)?;
        Ok(v)
    }
}

impl DecodeValue for u64 {
    fn decode_value(mut buf: &[u8]) -> eyre::Result<Self> {
        Ok(buf.read_u64_varint()?)
    }
}

impl EncodeValue for u64 {
    fn encode_value(self) -> eyre::Result<Vec<u8>> {
        let mut v = vec![];
        v.write_u64_varint(self)?;
        Ok(v)
    }
}

impl DecodeValue for bool {
    fn decode_value(mut buf: &[u8]) -> eyre::Result<Self> {
        Ok(buf.read_i32_varint()? != 0)
    }
}

impl EncodeValue for bool {
    fn encode_value(self) -> eyre::Result<Vec<u8>> {
        let mut v = vec![];
        v.write_i32_varint(self as i32)?;
        Ok(v)
    }
}

impl DecodeValue for Vec<u8> {
    fn decode_value(buf: &[u8]) -> eyre::Result<Self> {
        Ok(Vec::<u8>::decode(buf)?)
    }
}

impl EncodeValue for Vec<u8> {
    fn encode_value(self) -> eyre::Result<Vec<u8>> {
        Ok(self.encode_to_vec())
    }
}

impl DecodeValue for Orbit {
    fn decode_value(buf: &[u8]) -> eyre::Result<Self> {
        Ok(Orbit {
            id: u64::decode_value(buf)?,
        })
    }
}

impl EncodeValue for Orbit {
    fn encode_value(self) -> eyre::Result<Vec<u8>> {
        let mut v = vec![];
        v.write_u64_varint(self.id)?;
        Ok(v)
    }
}

impl DecodeValue for CelestialBody {
    fn decode_value(buf: &[u8]) -> eyre::Result<Self> {
        Ok(CelestialBody {
            id: u64::decode_value(buf)?,
        })
    }
}

impl EncodeValue for CelestialBody {
    fn encode_value(self) -> eyre::Result<Vec<u8>> {
        let mut v = vec![];
        v.write_u64_varint(self.id)?;
        Ok(v)
    }
}

impl DecodeValue for Status {
    fn decode_value(buf: &[u8]) -> eyre::Result<Self> {
        Ok(Status::decode(buf)?)
    }
}

impl DecodeValue for Vessel {
    fn decode_value(buf: &[u8]) -> eyre::Result<Self> {
        Ok(Self {
            id: u64::decode_value(buf)?,
        })
    }
}

impl EncodeValue for Vessel {
    fn encode_value(self) -> eyre::Result<Vec<u8>> {
        let mut v = vec![];
        v.write_u64_varint(self.id)?;
        Ok(v)
    }
}

impl DecodeValue for EditorShip {
    fn decode_value(buf: &[u8]) -> eyre::Result<Self> {
        Ok(Self {
            id: u64::decode_value(buf)?,
        })
    }
}

impl EncodeValue for EditorShip {
    fn encode_value(self) -> eyre::Result<Vec<u8>> {
        let mut v = vec![];
        v.write_u64_varint(self.id)?;
        Ok(v)
    }
}

impl DecodeValue for EditorParts {
    fn decode_value(buf: &[u8]) -> eyre::Result<Self> {
        Ok(Self {
            id: u64::decode_value(buf)?,
        })
    }
}

impl EncodeValue for EditorParts {
    fn encode_value(self) -> eyre::Result<Vec<u8>> {
        let mut v = vec![];
        v.write_u64_varint(self.id)?;
        Ok(v)
    }
}

impl DecodeValue for Parts {
    fn decode_value(buf: &[u8]) -> eyre::Result<Self> {
        Ok(Self {
            id: u64::decode_value(buf)?,
        })
    }
}

impl EncodeValue for Parts {
    fn encode_value(self) -> eyre::Result<Vec<u8>> {
        let mut v = vec![];
        v.write_u64_varint(self.id)?;
        Ok(v)
    }
}

impl DecodeValue for Part {
    fn decode_value(buf: &[u8]) -> eyre::Result<Self> {
        Ok(Self {
            id: u64::decode_value(buf)?,
        })
    }
}

impl EncodeValue for Part {
    fn encode_value(self) -> eyre::Result<Vec<u8>> {
        let mut v = vec![];
        v.write_u64_varint(self.id)?;
        Ok(v)
    }
}

impl DecodeValue for Editor {
    fn decode_value(buf: &[u8]) -> eyre::Result<Self> {
        Ok(Self {
            id: u64::decode_value(buf)?,
        })
    }
}

impl EncodeValue for Editor {
    fn encode_value(self) -> eyre::Result<Vec<u8>> {
        let mut v = vec![];
        v.write_u64_varint(self.id)?;
        Ok(v)
    }
}

impl DecodeValue for Decoupler {
    fn decode_value(buf: &[u8]) -> eyre::Result<Self> {
        Ok(Self {
            id: u64::decode_value(buf)?,
        })
    }
}

impl EncodeValue for Decoupler {
    fn encode_value(self) -> eyre::Result<Vec<u8>> {
        let mut v = vec![];
        v.write_u64_varint(self.id)?;
        Ok(v)
    }
}

impl DecodeValue for RODecoupler {
    fn decode_value(buf: &[u8]) -> eyre::Result<Self> {
        Ok(Self {
            id: u64::decode_value(buf)?,
        })
    }
}

impl EncodeValue for RODecoupler {
    fn encode_value(self) -> eyre::Result<Vec<u8>> {
        let mut v = vec![];
        v.write_u64_varint(self.id)?;
        Ok(v)
    }
}

impl DecodeValue for ReferenceFrame {
    fn decode_value(buf: &[u8]) -> eyre::Result<Self> {
        Ok(Self {
            id: u64::decode_value(buf)?,
        })
    }
}

impl EncodeValue for ReferenceFrame {
    fn encode_value(self) -> eyre::Result<Vec<u8>> {
        let mut v = vec![];
        v.write_u64_varint(self.id)?;
        Ok(v)
    }
}

impl DecodeValue for Resources {
    fn decode_value(buf: &[u8]) -> eyre::Result<Self> {
        Ok(Self {
            id: u64::decode_value(buf)?,
        })
    }
}

impl EncodeValue for Resources {
    fn encode_value(self) -> eyre::Result<Vec<u8>> {
        let mut v = vec![];
        v.write_u64_varint(self.id)?;
        Ok(v)
    }
}

impl DecodeValue for Resource {
    fn decode_value(buf: &[u8]) -> eyre::Result<Self> {
        Ok(Self {
            id: u64::decode_value(buf)?,
        })
    }
}

impl EncodeValue for Resource {
    fn encode_value(self) -> eyre::Result<Vec<u8>> {
        let mut v = vec![];
        v.write_u64_varint(self.id)?;
        Ok(v)
    }
}

impl DecodeValue for LaunchClamp {
    fn decode_value(buf: &[u8]) -> eyre::Result<Self> {
        Ok(Self {
            id: u64::decode_value(buf)?,
        })
    }
}

impl EncodeValue for LaunchClamp {
    fn encode_value(self) -> eyre::Result<Vec<u8>> {
        let mut v = vec![];
        v.write_u64_varint(self.id)?;
        Ok(v)
    }
}

impl DecodeValue for Module {
    fn decode_value(buf: &[u8]) -> eyre::Result<Self> {
        Ok(Self {
            id: u64::decode_value(buf)?,
        })
    }
}

impl EncodeValue for Module {
    fn encode_value(self) -> eyre::Result<Vec<u8>> {
        let mut v = vec![];
        v.write_u64_varint(self.id)?;
        Ok(v)
    }
}

impl DecodeValue for PFDecoupler {
    fn decode_value(buf: &[u8]) -> eyre::Result<Self> {
        Ok(Self {
            id: u64::decode_value(buf)?,
        })
    }
}

impl EncodeValue for PFDecoupler {
    fn encode_value(self) -> eyre::Result<Vec<u8>> {
        let mut v = vec![];
        v.write_u64_varint(self.id)?;
        Ok(v)
    }
}

File: new/KerbTk/src/KRPC/Encode.jl
module Encode

using ProtoBuf: ProtoBuf, ProtoDecoder, ProtoEncoder, encode, decode
using ProtoBuf.Codecs: _encode

using ..Schema: Schema

export encode_value, decode_value

function ProtoBuf.encode(x::T) where {T}
    buf = IOBuffer()
    encode(ProtoEncoder(buf), x)
    return take!(buf)
end

function ProtoBuf.decode(buf::Vector{UInt8}, t::Type{T})::T where {T}
    buf = IOBuffer(buf)
    return decode(ProtoDecoder(buf), t)
end

function decode_value(buf::Vector{UInt8}, ::Type{Dict{K,V}}) where {K,V}
    map = Dict{K,V}()
    raw = decode(buf, Schema.Dictionary)
    for entry in raw.entries
        key = decode_value(entry.key, K)
        value = decode_value(entry.value, V)
        map[key] = value
    end
    return map
end

function encode_value(dict::Dict{K,V}) where {K,V}
    entries = []
    for (key, value) in dict
        key = encode_value(key)
        value = encode_value(value)
        push!(entries, Schema.DictionaryEntry(key, value))
    end
    return encode(Schema.Dictionary(entries))
end

function decode_value(buf::Vector{UInt8}, ::Type{Vector{T}}) where {T}
    v = []
    raw = decode(buf, Schema.List)
    for item in raw.items
        push!(v, decode_value(item, T))
    end
    return v
end

function encode_value(v::Vector{T}) where {T}
    items = []
    for item in v
        push!(items, encode_value(item))
    end
    return encode(Schema.List(items))
end

function decode_value(buf::Vector{UInt8}, ::Type{Tuple{T,U,V}}) where {T,U,V}
    tuple = decode(buf, Schema.Tuple)
    return (
        decode_value(tuple.items[1], T),
        decode_value(tuple.items[2], U),
        decode_value(tuple.items[3], V),
    )
end

function encode_value(x::Tuple{T,U,V}) where {T,U,V}
    v = []
    push!(v, encode_value(x[1]))
    push!(v, encode_value(x[2]))
    push!(v, encode_value(x[3]))
    return encode(Schema.Tuple(v))
end

function decode_value(buf::Vector{UInt8}, ::Type{Tuple{T,U,V,W}}) where {T,U,V,W}
    tuple = decode(buf, Schema.Tuple)
    return (
        decode_value(tuple.items[1], T),
        decode_value(tuple.items[2], U),
        decode_value(tuple.items[3], V),
        decode_value(tuple.items[4], W),
    )
end

function encode_value(x::Tuple{T,U,V,W}) where {T,U,V,W}
    v = []
    push!(v, encode_value(x[1]))
    push!(v, encode_value(x[2]))
    push!(v, encode_value(x[3]))
    push!(v, encode_value(x[4]))
    return encode(Schema.Tuple(v))
end

function decode_value(buf::Vector{UInt8}, ::Type{Tuple{T,U,V,W,X}}) where {T,U,V,W,X}
    tuple = decode(buf, Schema.Tuple)
    return (
        decode_value(tuple.items[1], T),
        decode_value(tuple.items[2], U),
        decode_value(tuple.items[3], V),
        decode_value(tuple.items[4], W),
        decode_value(tuple.items[5], X),
    )
end

function encode_value(x::Tuple{T,U,V,W,X}) where {T,U,V,W,X}
    v = []
    push!(v, encode_value(x[1]))
    push!(v, encode_value(x[2]))
    push!(v, encode_value(x[3]))
    push!(v, encode_value(x[4]))
    push!(v, encode_value(x[5]))
    return encode(Schema.Tuple(v))
end

decode_value(::Vector{UInt8}, ::Type{Tuple{}}) = ()
encode_value(::Type{Tuple{}}) = []

function decode_value(buf::Vector{UInt8}, ::Type{String})
    buf = IOBuffer(buf)
    len = decode(ProtoDecoder(buf), UInt64)
    pos = position(buf) + 1
    buf = take!(buf)[pos:end]
    @assert length(buf) == len
    return String(buf)
end

function encode_value(s::String)
    v = IOBuffer()
    _encode(v, UInt64(length(v)))
    write(v, s)
    return take!(v)
end

function decode_value(buf::Vector{UInt8}, ::Type{Union{T,Nothing}}) where {T}
    if buf == [0x0]
        return nothing
    else
        return decode_value(buf, T)
    end
end

function encode_value(x::Union{T,Nothing}) where {T}
    if isnothing(x)
        return [0x0]
    else
        return encode_value(convert(T, x))
    end
end

function decode_value(
    buf::Vector{UInt8}, ::Type{T}
) where {T<:Union{Float32,Float64,UInt32,UInt64,Int32}}
    return decode(ProtoDecoder(IOBuffer(buf)), T)
end

function encode_value(x::T) where {T<:Union{Float32,Float64,UInt32,UInt64,Int32}}
    buf = IOBuffer()
    _encode(buf, x)
    return take!(buf)
end

function decode_value(buf::Vector{UInt8}, ::Type{Bool})
    return decode(ProtoDecoder(IOBuffer(buf)), Int32) != 0
end

function encode_value(x::Bool)
    buf = IOBuffer()
    _encode(buf, Int32(x ? 1 : 0))
    return take!(buf)
end

function decode_value(buf::Vector{UInt8}, ::Type{Vector{UInt8}})
    return decode(ProtoDecoder(IOBuffer(buf)), Vector{UInt8})
end

function encode_value(x::Vector{UInt8})
    buf = IOBuffer()
    encode(ProtoEncoder(buf), x)
    return take!(buf)
end

function decode_value(buf::Vector{UInt8}, ::Type{Schema.Status})
    return decode(ProtoDecoder(IOBuffer(buf)), Schema.Status)
end

using ..KRPC: CelestialBody, AbstractKRPCValue

function encode_value(x::T) where {T<:AbstractKRPCValue}
    buf = IOBuffer()
    _encode(buf, UInt64(x.inner))
    return take!(buf)
end

function decode_value(buf::Vector{UInt8}, ::Type{T}) where {T<:AbstractKRPCValue}
    return T(decode(ProtoDecoder(IOBuffer(buf)), UInt64))
end

end
